modules/rating/src/main/glicko2/Result.scala

```diff
@@ -2,50 +2,49 @@ package lila.rating.glicko2
 
 trait Result:
 
-  def getScore(player: Rating): Double
+  val first: Rating
+  val second: Rating
 
-  def getOpponent(player: Rating): Rating
+  def getAdvantage(advantage: Double, player: Rating): Double =
+    if player == first then advantage / 2.0d else -advantage / 2.0d
 
-  def participated(player: Rating): Boolean
+  def getScore(player: Rating): Double
 
-  def players: List[Rating]
+  def getOpponent(player: Rating): Rating
 
-// score from 0 (opponent wins) to 1 (player wins)
-class FloatingResult(player: Rating, opponent: Rating, score: Float) extends Result:
+  def participated(player: Rating): Boolean = player == first || player == second
 
-  def getScore(p: Rating) = if p == player then score else 1 - score
+  def players: List[Rating] = List(first, second)
 
-  def getOpponent(p: Rating) = if p == player then opponent else player
+class BinaryResult(val first: Rating, val second: Rating, score: Boolean) extends Result:
+  private val POINTS_FOR_WIN  = 1.0d
+  private val POINTS_FOR_LOSS = 0.0d
 
-  def participated(p: Rating) = p == player || p == opponent
+  def getScore(p: Rating) = if p == first then if score then POINTS_FOR_WIN else POINTS_FOR_LOSS
+  else if score then POINTS_FOR_LOSS
+  else POINTS_FOR_WIN
 
-  def players = List(player, opponent)
+  def getOpponent(p: Rating) = if p == first then second else first
 
-final class GameResult(winner: Rating, loser: Rating, isDraw: Boolean) extends Result:
+final class GameResult(val first: Rating, val second: Rating, outcome: Option[Boolean]) extends Result:
```

Comment:
I'm concerned that you changed the semantics of this class while keeping typing compatibility.

`winner: Rating, loser: Rating` and `val first: Rating, val second: Rating` have the same types, but mean different things. Code using it might break without the compiler noticing.

Reply:
Good point.  I changed the only other consumer of this class (`PuzzleFinisher.scala`) to consume `BinaryResult` instead.

I tried to think of a better name than `GameResult` to describe this concept... I needed a way to identify which player had the first move advantage.